### ISSUE TITLE
What is the best way to upsert a collection ?

### ISSUE THREAD

**Issue by dalanad** [mikro-orm/mikro-orm#1042]
In the following scenario a whole `SalesOrder` entity is passed in to the server
items collection of sales order entity can have new items, removed items and updated items.
just persisting the `SalesOrder` don't work (wont remove or update)
this is the way i got it to work. is there a better way? or an I missing something?
```
async save(@Body() item: SalesOrder) {

let result: SalesOrder;
result = await em.getRepository<SalesOrder>(SalesOrder).findOne(item.id);
item.items.forEach(k => k.salesOrder = item.id)
await this.em.getRepository(SalesOrderItem).nativeDelete({ salesOrder: item.id })
for (const [i, it] of item.items.entries()) {
    const x = em.getRepository(SalesOrderItem).create({ ...it })
    await em.persistAndFlush(x)
}
wrap(result).assign({ ...item, items: null }, {})

await em.getRepository(SalesOrder).persistAndFlush(result);
await em.flush()
}
```

```
@Entity()
export class SalesOrder extends BaseEntity {
    @PrimaryKey( )
    id: number;

    @OneToMany(i => SalesOrderItem, t => t.salesOrder, { eager: true, cascade: [Cascade.ALL] })
    items = new Collection<SalesOrderItem>(this); 
}

@Entity()
export class SalesOrderItem extends BaseEntity {

    @ManyToOne({ entity: () => InventoryItem, eager: true, primary: true })
    item: IdentifiedReference<InventoryItem>;

    @Property({ type: JsonType, columnType: 'jsonb', nullable: true })
    configuration?: SalesOrderItem[]

    @ManyToOne({ entity: () => SalesOrder, hidden: true, primary: true })
    salesOrder: IdentifiedReference<SalesOrder>;

    @Property({ nullable: false })
    qty: number;

    @Property({ nullable: false })
    unitPrice: number;

    [PrimaryKeyType]: [string, string]
}
```

**Comment by B4nan** [comment]
Flushing in for of loop is in general anti-pattern and is almost certainly not needed. Your code looks very weird and very imperformant, e.g. `const x = em.getRepository(SalesOrderItem).create({ ...it })` is wrong, why would you reconstruct an entity that is already entity? That will break a lot of internal things, from managed `it` entity (that would produce update query) you are creating new entity that will produce insert query. If this code works for you, I am really surprised :P 

Another issue is here: `item.items.forEach(k => k.salesOrder = item.id)`. You are assigning `item.id` (so the PK) to `salesOrder` which is an entity, not scalar. Or is the parameter a DTO and you are just typing it wrongly? 

Lastly your approach will 1. fire delete query outside of any transaction, 2. create N small transactions for each item in the for loop, and 3. create one more transaction for the last flush you do. This is extremely unsafe way, even if it works for you locally, you should do everything in a single transaction, I would be very scared to have such code in production.

I don't really understand what should "upserting a collection mean". The term feels very weird to me, you may upsert a single entity/db row, not a collection - collection is a virtual structure (especially the 1:m collection). Instead of working with such collection (which is by definition inverse side), you should try to work with the owning side (where the m:1 is defined). In the end that is what you already doing in your code (but in a very weird way).

Btw take a look at orpah removal, you might be interested in that if you want items removed from 1:m to produce delete query. 

https://mikro-orm.io/docs/cascading/#orphan-removal

Otherwise removing item from such collection will produce update query that will set the relation to null (which is not allowed in your entity definition as the m:1 property is not nullable).

I can't help much with this, as I don't really understand what you are up to and there are too many details missing (as well as I feel like I would be doing your work for free, tbh).

**Comment by dalanad** [comment]
Very Sorry for the trouble.
I went through the source code and
Initial issues led to this seem to be solved with d40fcfa772efa9f84484293a3b24da1cbd085add .
Also there were a problem with composite p keys  

Remaining is that I need the items of the Collection also updated when using `wrap(entity).assign()`
I've put a loop to assign each item of the collection. (this collection would not have a large number of elements)

for simplicity of understanding, in `assignCollection` function of  `EntityAssigner`
const items = value.map((item) => this.createCollectionItem(item, em, prop, invalid, options)  **.assign(item)**   );
adding `.assign(item)` solves the case

I'll let know if I find  anything
